### PR TITLE
plugin/kubernetes: implement HasSynced() 

### LINKED ISSUE #1006: mw/kubernetes: block until ready?
The kubernetes tests have this:
~~~
	// Work-around for timing condition that results in no-data being returned in test environment.
	time.Sleep(3 * time.Second)
~~~
Which means that even in our code we need a workaround. This should be replaced with either a call to a blocking channel that gets closed once we've fully started or something else.

This is also needed for #985

### PR DESCRIPTION
**Pull request by miekg**
### 1. What does this pull request do?

Implements HasSynced() that holds startup until the API has fully synced. Removing the race in the startup we now have.

Fixes #1006

### CODE REVIEW

**Comment by fturib** [comment]
Hum.. this PR includes change for optimization on endPoint list index ... @rajansandeep : can you check that this change has no side-effect ?

**Comment by miekg** [comment]
@fturib  you mean this commit: https://github.com/coredns/coredns/pull/1155/commits/ec84ff2d99ebd47c9fc2917982927a29e83f6a51 ?

**Comment by rajansandeep** [comment]
@fturib, the change is a cleanup and has no side effect.

**Comment by miekg** [comment]
hmmm, keep being flaky
~~~
-- FAIL: TestKubernetes (0.23s)
	helpers.go:115: rcode is "NXDOMAIN", expected "NOERROR"
	helpers.go:291: ;; opcode: QUERY, status: NXDOMAIN, id: 62757
	;; flags: qr aa rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 0
	;; QUESTION SECTION:
	;*.svc-1-a.test-1.svc.cluster.local.	IN	SRV
	;; AUTHORITY SECTION:
	cluster.local.	300	IN	SOA	ns.dns.cluster.local. hostmaster.cluster.local. 1508510551 7200 1800 86400 60
~~~
faster flaky though

**Comment by miekg** [comment]
breaks on different questions, which makes sense because the the `range` we use in the tests. So even with HasSynced() we need to wait for a bit?
~~~
-- FAIL: TestKubernetes (0.21s)
	helpers.go:115: rcode is "NXDOMAIN", expected "NOERROR"
	helpers.go:291: ;; opcode: QUERY, status: NXDOMAIN, id: 7853
	;; flags: qr aa rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 0
	;; QUESTION SECTION:
	;headless-svc.test-1.svc.cluster.local.	IN	A
	;; AUTHORITY SECTION:
	cluster.local.	300	IN	SOA	ns.dns.cluster.local. hostmaster.cluster.local. 1508513248 7200 1800 86400 60
=== RUN TestMetricsServer
~~~

**Comment by miekg** [comment]
believe this is still a good PR to merge, not complete OK with the sleep(1s) that is still needed, but an improvement over the previous behavior.

**Comment by fturib** [comment]
@miekg : we reviewed the change about endPointList() in findServices() of kubernetes.go. And would advise to revert your commit : it was somehow wrong already, and your change will make another wrong doing. I am opening an issue to take care of this findService() issue.

**Comment by miekg** [comment]
Ok. sgtm.

Why about the HasSynced change?

**Comment by miekg** [comment]
Note: nothing has been commited, you mean we should revert somethting?

**Comment by fturib** [comment]
I am referring only to the changes in 9  plugin/kubernetes/kubernetes.go. it is about the optimization in findService()  (or cleanup), and seems not linked to HasSynced() feature.

I am not sure about the wording.. but I see that change in the commit https://github.com/coredns/coredns/pull/1155/commits/da0c8edced16dc82c1dfa4bea4d8509bf7883e86, although it is not merged.

**Comment by miekg** [comment]
I don't fully understand. If it was wrong already, can you send a PR on how it should look.

The change I made in this PR is a noop - so I'm not sure what you're referring at

**Comment by fturib** [comment]
I am doing the change in same time locally. 
Eventually, you are right,  it is a noop, only impacting efficiency. 
So .. finally you can keep. 

I am doing the change locally, but needs to run the Tests, and then find how to setup a PR. 
It will be my first .. so will take some time :)

**Comment by miekg** [comment]
Ok, no problem.

So what about this pr? Merge? Or back out and wait for yours?
(we're "only" pushing to master so we can play around a bit :))

**Comment by fturib** [comment]
For me you can merge.

**Comment by fturib** [comment]
I opened an issue here for my concern on endPointList : #1156

**Comment by fturib** [comment]
hum .. you may think I am a little instable, but looking into the code you propose here, I think you made a 'maybe' infinite loop in the Setup function.

=> you may want to break the wait of readiness after an amount of time or loops - and raise an Error is not HasSynced().

File: plugin/kubernetes/setup.go [hunk 1]
```diff
@@ -39,8 +39,14 @@ func setup(c *caddy.Controller) error {
 	c.OnStartup(func() error {
 		go kubernetes.APIConn.Run()
 		if kubernetes.APIProxy != nil {
-			go kubernetes.APIProxy.Run()
+			kubernetes.APIProxy.Run()
 		}
+		synced := false
+		for synced == false {
```
Comment (aledbf) [inline plugin/kubernetes/setup.go:45]: Please check https://github.com/kubernetes/client-go/blob/master/tools/cache/shared_informer.go#L104

You can use
```
        if !cache.WaitForCacheSync(stopCh,
		kubernetes.APIConn.HasSynced,
	) {
		return fmt.Errorf("Timed out waiting for caches to sync")
	}
```
from package `k8s.io/client-go/tools/cache`

Basically this waits until the cache has synced.

Comment (miekg) [reply plugin/kubernetes/setup.go:45]: Interesting. I'll check tomorrow - might still merge this and then follow up.

**Pull request merged**